tools: move eslint and install babel-eslint

targos opened the pull request:
Commit 1:

    tools: move eslint from tools to tools/node_modules

    This is required because we need to add the babel-eslint dependency
    and it has to be able to resolve "eslint".

    Refs: https://github.com/nodejs/node/pull/17755

Commit 2:

    tools: add babel-eslint

    Create tools/update-babel-eslint.sh script and execute it to do the
    first installation of the package.
    Update tools/license-builder.sh and execute it to add babel-eslint's
    license to our LICENSE file.

Commit 3:

    tools: use babel-eslint as ESLint parser

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
tools

/cc @mcollina

MylesBorins commented:
@targos how many files does this create? I'm not going to block, but we should be consistent about decision making on new dependencies

@refack have you had any progress on new ways to vendor deps?

targos commented:
This creates 1217 files.

mcollina commented:
Commit 2 and 3 could be reverted after AsyncIterators are added to ESlint itself.

maclover7 commented:
> how many files does this create? I'm not going to block, but we should be consistent about decision making on new dependencies

Agree with @MylesBorins's comment. Don't want to start a big bikeshed on this PR, but _in theory_ if this lands, so should #17320, which is very similar in nature.

Trott commented:
Why do we want to add babel-eslint as the parser? (Not objecting, of course, but I do want to understand the motivation!)

mcollina commented:
To support async-iterators. They are curreny a stage 3 proposal, and it’s not supported into eslint yet.

benjamingr commented:
@sebmck

cjihrig commented:
According to https://twitter.com/geteslint/status/944829414159810560, there is an issue with babel-eslint and eslint 1.14.0. The issue appears fixed with babel-eslint 8.1.0. Assuming that we'll eventually update to 1.14.0+, it might make sense to update to both of those versions now.

I'm also not big on adding this to the repo, although I won't block this either. Even if we remove them later, they will still be part of the git history.

targos commented:
I also don't like having to add this dependency, but do we have an alternative?

This will be necessary for `import.meta` as well.

BTW 1054 of the new files come from `lodash` :(

mcollina commented:
> I also don't like having to add this dependency, but do we have an alternative?

No, if we want to experiment with the new language feature in node core _before_ they are finalyzed.

Is there anyone objecting landing this?

gibfahn commented:
>I also don't like having to add this dependency, but do we have an alternative?

Would git submodules work here?

targos commented:
Added a commit to update ESLint to 4.14.0 and updated babel-eslint to the latest version.

CI: https://ci.nodejs.org/job/node-test-pull-request/12375/

benjamingr commented:
Ping @targos

mcollina commented:
Are we ok to land this? cc @nodejs/tsc
@targos can you get it rebased?

targos commented:
It looks like the merge of master is making CI unhappy. Let me rebase this properly.

targos commented:
Updated

benjamingr commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/12459/console

targos commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/12460/

benjamingr commented:
@targos the older CI I started looks good from what I can tell, everything that is unstable is also unstable on other builds unrelated to the PR.

targos commented:
@benjamingr  Oh, sorry my tab was not up-to-date. I didn't see your message. I stopped my run.

mcollina commented:
CI (again, problems with Windows): https://ci.nodejs.org/job/node-test-pull-request/12480/

targos commented:
Landed in 3dc30632755713179f345f4af024bd904c6162d0, 7a52c51e811cd59db0d2b1b6d6b8b2e7f917d65c, b043a70201587920292512314b8815d33f62083f and 4d96c17e054fd4735063dc4aa4341ea76ca12b49

evanlucas commented:
Does this need to land on v9.x?

MylesBorins commented:
Should we backport this to v6.x or v8.x?

BethGriggs commented:
If this gets backported it should come with https://github.com/nodejs/node/pull/19287
